decred/decred/dcr/addrlib.py

```diff
@@ -0,0 +1,447 @@
+        """
+        A copy of the scriptHash.
+
+        Returns:
+            ByteArray: The hash.
+        """
+        return self.scriptHash.copy()
+
+
+class AddressEdwardsPubKey(Address):
+    """unimplemented"""
+
+    def __init__(*a, **k):
+        raise NotImplementedError("AddressEdwardsPubKey not implemented")
+
+
+class AddressSecSchnorrPubKey(Address):
+    """unimplemented"""
+
+    def __init__(*a, **k):
+        raise NotImplementedError("AddressSecSchnorrPubKey implemented")
+
+
+def encodeAddress(netID, k):
+    """
+    Base-58 encode the number, with the netID prepended byte-wise.
+
+    Args:
+        netID (byte-like): The addresses network encoding ID.
+        k (ByteArray): The pubkey or pubkey-hash or script-hash.
+
+    Returns:
+        string: Base-58 encoded address.
+    """
+    b = ByteArray(netID)
+    b += k
+    b += checksum(b.b)
+    return b58encode(b.bytes()).decode()
+
+
+def decodeAddress(addr, netParams):
+    """
+    DecodeAddress decodes the base-58 encoded address and returns the Address if
+    it is a valid encoding for a known address type and is for the provided
+    network.
+
+    Args:
+        netParams (module): The network parameters.
+    """
+    # Switch on decoded length to determine the type.
+    decoded, netID = b58CheckDecode(addr)
+    if netID == netParams.PubKeyAddrID:
+        return decodeAddressPubKey(decoded, netParams)
+    elif netID == netParams.PubKeyHashAddrID:
+        return AddressPubKeyHash(decoded, netParams, STEcdsaSecp256k1)
+    elif netID == netParams.PKHEdwardsAddrID:
+        raise NotImplementedError("Edwards signatures not implemented")
+    elif netID == netParams.PKHSchnorrAddrID:
+        raise NotImplementedError("Schnorr signatures not implemented")
+    elif netID == netParams.ScriptHashAddrID:
+        return AddressScriptHash(decoded, netParams)
+    raise NotImplementedError(f"unknown network ID {netID}")
+
+
+def decodeAddressPubKey(decoded, netParams):
+    """
+    decodeAddressPubKey decodes a pubkey-type address from the serialized
+    pubkey.
+
+    Args:
+        decoded (bytes): A 33 bytes decoded pubkey such as would be decoded
+            from a base58 string. The first byte indicates the signature suite.
+            For compressed secp256k1 pubkeys, use AddressSecpPubKey directly.
+        netParams (module): The network parameters.
+    """
+    if len(decoded) != 33:
+        raise NotImplementedError(f"unable to decode pubkey of length {len(decoded)}")
+    # First byte is the signature suite and ybit.
+    suite = decoded[0]
+    suite &= 127
+    ybit = not (decoded[0] & (1 << 7) == 0)
+    toAppend = 0x02
+    if ybit:
+        toAppend = 0x03
+
+    if suite == STEcdsaSecp256k1:
+        b = ByteArray(toAppend) + decoded[1:]
+        return AddressSecpPubKey(b, netParams)
+        # return NewAddressEdwardsPubKey(decoded, netParams)
+        raise NotImplementedError("Edwards signatures not implemented")
```

Comment:
The commented out line can go, and the unreachable `raise` line should probably go too.

Reply:
Whoops. Looks like I lost a line here.